File: Rude-Chat-flask-agents/MCP-Server/MCPServer/Services/Agents/CoreAgent.cs
using MCPServer.Models;

namespace MCPServer.Services.Agents;

/// <summary>
/// Core agent for basic system tools and general-purpose functionality.
/// Handles tools that don't belong to specific domain areas.
/// </summary>
public class CoreAgent : BaseAgent
{
    public CoreAgent(
        ILogger<CoreAgent> logger, 
        IConfiguration configuration) 
        : base(logger, configuration)
    {
    }

    public override string AgentId => "core-agent";
    public override string Name => "Core Agent";
    public override string Description => "Core agent for basic system tools and general-purpose functionality";
    public override IEnumerable<string> Domains => new[] { "core", "system", "basic", "general" };

    protected override Task OnInitializeAsync(string? userToken)
    {
        _logger.LogInformation("Initializing Core Agent");
        return Task.CompletedTask;
    }

    public override Task<IEnumerable<McpTool>> GetAvailableToolsAsync()
    {
        try
        {
            var tools = new List<McpTool>
            {
                new McpTool
                {
                    Name = "hello_world",
                    Description = "A simple Hello World tool that greets the user",
                    InputSchema = new McpToolInputSchema
                    {
                        Type = "object",
                        Properties = new Dictionary<string, McpProperty>
                        {
                            ["name"] = new McpProperty
                            {
                                Type = "string",
                                Description = "The name to greet"
                            }
                        },
                        Required = new[] { "name" }
                    }
                },
                new McpTool
                {
                    Name = "system_info",
                    Description = "Get information about the MCP server system and available agents",
                    InputSchema = new McpToolInputSchema
                    {
                        Type = "object",
                        Properties = new Dictionary<string, McpProperty>(),
                        Required = Array.Empty<string>()
                    }
                }
            };

            _logger.LogDebug("Core Agent providing {ToolCount} tools", tools.Count);
            return Task.FromResult<IEnumerable<McpTool>>(tools);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Failed to get Core tools");
            return Task.FromResult(Enumerable.Empty<McpTool>());
        }
    }

    public override Task<McpToolCallResponse> ExecuteToolAsync(McpToolCallRequest request)
    {
        _logger.LogInformation("Core Agent executing tool: {ToolName}", request.Name);

        try
        {
            return request.Name switch
            {
                "hello_world" => ExecuteHelloWorldAsync(request.Arguments),
                "system_info" => ExecuteSystemInfoAsync(request.Arguments),
                _ => Task.FromResult(CreateErrorResponse($"Unknown core tool: {request.Name}"))
            };
        }
        catch (Exception ex)
        {
            return Task.FromResult(CreateErrorResponse($"Core Agent failed to execute tool {request.Name}", ex));
        }
    }

    public override Task<bool> CanHandleToolAsync(string toolName)
    {
        return Task.FromResult(IsCoreTool(toolName));
    }

    protected override Task<bool> PerformHealthCheckAsync()
    {
        // Core agent is always healthy as it doesn't depend on external services
        return Task.FromResult(true);
    }

    protected override Task<Dictionary<string, object>> GetHealthMetadataAsync()
    {
        var metadata = new Dictionary<string, object>
        {
            ["service_type"] = "Core System",
            ["version"] = "1.0.0",
            ["startup_time"] = DateTime.UtcNow.ToString("yyyy-MM-dd HH:mm:ss UTC")
        };

        return Task.FromResult(metadata);
    }

    private Task<McpToolCallResponse> ExecuteHelloWorldAsync(Dictionary<string, object> arguments)
    {
        var name = arguments.GetValueOrDefault("name", "World")?.ToString() ?? "World";
        
        var response = CreateSuccessResponse($"Hello, {name}! This is a greeting from the Azure MCP Server Core Agent.");
        return Task.FromResult(response);
    }

    private Task<McpToolCallResponse> ExecuteSystemInfoAsync(Dictionary<string, object> arguments)
    {
        var systemInfo = "🤖 **Azure MCP Server System Information**\n\n";
        systemInfo += "**Architecture:** Agent-based modular design\n";
        systemInfo += "**Available Agent Domains:**\n";
        systemInfo += "• Core - Basic system tools\n";
        systemInfo += "• ADX - Azure Data Explorer operations\n";
        systemInfo += "• Maps - Azure Maps and location services\n";
        systemInfo += "• Documents - Document management and search\n";
        systemInfo += "• Resources - Azure resource management\n\n";
        systemInfo += "**Features:**\n";
        systemInfo += "• User-based authentication with On-Behalf-Of flow\n";
        systemInfo += "• Dynamic tool execution with unlimited chaining\n";
        systemInfo += "• Agent health monitoring\n";
        systemInfo += "• Extensible architecture for new domains\n\n";
        systemInfo += "Use agent-specific tools or ask for suggestions to get started!";

        var response = CreateSuccessResponse(systemInfo);
        return Task.FromResult(response);
    }

    private static bool IsCoreTool(string toolName)
    {
        return toolName switch
        {
            "hello_world" => true,
            "system_info" => true,
            _ => false
        };
    }
}

File: Rude-Chat-flask-agents/Python-MCP-Server/src/agents/core_agent.py
"""
Core agent for basic system tools and general-purpose functionality.
Handles tools that don't belong to specific domain areas.
"""

import logging
from typing import List, Dict, Any, Optional
from datetime import datetime

from .base_agent import BaseAgent
try:
    from ..models.mcp_models import McpTool, McpToolCallRequest, McpToolCallResponse, McpProperty, McpToolInputSchema
except ImportError:
    # Fallback for when running directly
    import sys
    import os
    sys.path.append(os.path.dirname(os.path.dirname(__file__)))
    from models.mcp_models import McpTool, McpToolCallRequest, McpToolCallResponse, McpProperty, McpToolInputSchema

logger = logging.getLogger(__name__)

class CoreAgent(BaseAgent):
    """Core agent for basic system tools and general-purpose functionality"""
    
    def __init__(self, config: Dict[str, Any]):
        super().__init__(config)
        
    @property
    def agent_id(self) -> str:
        return "core-agent"
    
    @property
    def name(self) -> str:
        return "Core Agent"
    
    @property
    def description(self) -> str:
        return "Core agent for basic system tools and general-purpose functionality"
    
    @property
    def domains(self) -> List[str]:
        return ["core", "system", "basic", "general"]
    
    async def _on_initialize_async(self, user_token: Optional[str]) -> None:
        """Initialize the Core Agent"""
        logger.info("Initializing Core Agent")
        # No specific initialization needed for core agent
    
    async def get_available_tools_async(self) -> List[McpTool]:
        """Get all tools that this agent can execute"""
        try:
            tools = [
                McpTool(
                    name="hello_world",
                    description="A simple Hello World tool that greets the user",
                    input_schema=McpToolInputSchema(
                        type="object",
                        properties={
                            "name": McpProperty(
                                type="string",
                                description="The name to greet"
                            )
                        },
                        required=["name"]
                    )
                ),
                McpTool(
                    name="system_info",
                    description="Get information about the MCP server system and available agents",
                    input_schema=McpToolInputSchema(
                        type="object",
                        properties={},
                        required=[]
                    )
                ),
                McpTool(
                    name="echo",
                    description="Echo back the provided message",
                    input_schema=McpToolInputSchema(
                        type="object",
                        properties={
                            "message": McpProperty(
                                type="string",
                                description="The message to echo back"
                            )
                        },
                        required=["message"]
                    )
                ),
                McpTool(
                    name="get_time",
                    description="Get the current server time",
                    input_schema=McpToolInputSchema(
                        type="object",
                        properties={},
                        required=[]
                    )
                )
            ]
            
            logger.debug(f"Core Agent providing {len(tools)} tools")
            return tools
        except Exception as e:
            logger.error(f"Failed to get Core tools: {str(e)}")
            return []
    
    async def execute_tool_async(self, request: McpToolCallRequest) -> McpToolCallResponse:
        """Execute a tool request"""
        logger.info(f"Core Agent executing tool: {request.name}")
        
        try:
            if request.name == "hello_world":
                return await self._execute_hello_world(request.arguments)
            elif request.name == "system_info":
                return await self._execute_system_info(request.arguments)
            elif request.name == "echo":
                return await self._execute_echo(request.arguments)
            elif request.name == "get_time":
                return await self._execute_get_time(request.arguments)
            else:
                return self._create_cannot_answer_response(f"Unknown core tool: {request.name}")
        except Exception as e:
            return self._create_error_response(f"Core Agent failed to execute tool {request.name}", e)
    
    async def can_handle_tool_async(self, tool_name: str) -> bool:
        """Check if this agent can handle the specified tool"""
        core_tools = ["hello_world", "system_info", "echo", "get_time"]
        return tool_name in core_tools
    
    async def _perform_health_check_async(self) -> bool:
        """Perform Core Agent health check"""
        try:
            # Simple health check - verify we can get tools
            tools = await self.get_available_tools_async()
            return len(tools) > 0
        except Exception as e:
            logger.warning(f"Core Agent health check failed: {str(e)}")
            return False
    
    async def _execute_hello_world(self, arguments: Dict[str, Any]) -> McpToolCallResponse:
        """Execute the hello_world tool"""
        name = arguments.get("name", "World")
        message = f"Hello, {name}! This is the Core Agent from the Python MCP Server."
        return McpToolCallResponse.success(message)
    
    async def _execute_system_info(self, arguments: Dict[str, Any]) -> McpToolCallResponse:
        """Execute the system_info tool"""
        info = {
            "server_name": "Python Azure MCP Server",
            "version": "1.0.0",
            "agent_id": self.agent_id,
            "agent_name": self.name,
            "timestamp": datetime.utcnow().isoformat(),
            "domains": self.domains
        }
        
        message = f"MCP Server System Information:\n"
        for key, value in info.items():
            if isinstance(value, list):
                message += f"- {key}: {', '.join(value)}\n"
            else:
                message += f"- {key}: {value}\n"
        
        return McpToolCallResponse.success(message)
    
    async def _execute_echo(self, arguments: Dict[str, Any]) -> McpToolCallResponse:
        """Execute the echo tool"""
        message = arguments.get("message", "")
        if not message:
            return self._create_error_response("No message provided to echo")
        
        response = f"Echo: {message}"
        return McpToolCallResponse.success(response)
    
    async def _execute_get_time(self, arguments: Dict[str, Any]) -> McpToolCallResponse:
        """Execute the get_time tool"""
        current_time = datetime.utcnow().isoformat()
        message = f"Current server time (UTC): {current_time}"
        return McpToolCallResponse.success(message)
